infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/runtime/LogReplicationSourceBaseRouter.java

```diff
@@ -0,0 +1,473 @@
+package org.corfudb.infrastructure.logreplication.runtime;
+
+import lombok.Getter;
+import lombok.Setter;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.infrastructure.LogReplicationRuntimeParameters;
+import org.corfudb.infrastructure.logreplication.infrastructure.ClusterDescriptor;
+import org.corfudb.infrastructure.logreplication.infrastructure.CorfuReplicationManager;
+import org.corfudb.infrastructure.logreplication.runtime.fsm.LogReplicationRuntimeEvent;
+import org.corfudb.infrastructure.logreplication.transport.client.ChannelAdapterException;
+import org.corfudb.infrastructure.logreplication.transport.client.IClientChannelAdapter;
+import org.corfudb.infrastructure.logreplication.transport.server.IServerChannelAdapter;
+import org.corfudb.protocols.service.CorfuProtocolMessage;
+import org.corfudb.runtime.LogReplication.LogReplicationSession;
+import org.corfudb.runtime.clients.IClient;
+import org.corfudb.runtime.clients.IClientRouter;
+import org.corfudb.runtime.exceptions.NetworkException;
+import org.corfudb.runtime.exceptions.unrecoverable.UnrecoverableCorfuInterruptedError;
+import org.corfudb.runtime.proto.RpcCommon;
+import org.corfudb.runtime.proto.service.CorfuMessage;
+import org.corfudb.util.CFUtils;
+
+import javax.annotation.Nonnull;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static org.corfudb.protocols.CorfuProtocolCommon.getUuidMsg;
+import static org.corfudb.protocols.service.CorfuProtocolMessage.getDefaultProtocolVersionMsg;
+import static org.corfudb.protocols.service.CorfuProtocolMessage.getRequestMsg;
+
+@Slf4j
+public class LogReplicationSourceBaseRouter implements IClientRouter {
+
```

Comment:
Furthermore, I still didn't see the reason or benefits to give each session a "router", especially they are using the same underlying gRPC channels right? From a quick examine of the code, the session in these router are used for 
* Set header msg session field
* Pass down to gRPC channel adapter

then why can't we just expose session as an argument? also, some parts are common and not associated with a session, like `sendResponse` in `LogReplicationSourceServerRouter`, the header is already set before invoke of this method

Reply:
yes, It is possible. But we would have a heavy use of locks in this layer then..
So I opted for having a router per session. Another advantage of this would be that serializing request/responses would be pushed a layer lower. So, the only part that would be serialized is the actual sending of requests.

We can discuss this more offline.